make update_scope recursive while keeping trace_gadt_instances

garrigue opened the pull request:
This is a new version #9815, which does not try to remove `trace_gadt_instances` simultaneously.

garrigue committed:
make update_scope recursive while keeping trace_gadt_instances

garrigue committed:
add log

garrigue committed:
accept tests/let-syntax

garrigue committed:
forget scope when expanding trace

lpw25 commented on typing/ctype.ml:
```diff
@@ -1756,8 +1760,20 @@ let enforce_constraints env ty =
    Also expand #-types. *)
-let full_expand env ty =
-  let ty = repr (expand_head env ty) in
+let full_expand ~may_forget_scope env ty =
```
I'm not sure that the changes to `full_expand` are the right behaviour. I think that a node whose level is outside the scope of the equation should probably be printed as `type a` rather than `type a = foo` since it would not be expanded without producing ambiguity.

garrigue replied:
If the error came from this expansion, it seems better to show it, otherwise the message is rather confusing.
Maybe I could try to refine this behavior, so that we can explain in the error message that this (probably expected) expansion is not available here.

lpw25 commented on typing/typecore.ml:
```diff
@@ -4263,9 +4264,9 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
-    {desc = Tarrow(Nolabel,ty_arg,ty_res,_); level = lv}
-    when is_inferred sarg ->
+  let lv = (repr ty_expected').level in
+  match expand_head env (correct_levels ty_expected') with
```
I think this change raises a number of issues:
1. There are many other uses of `expand_head`, quite a few of which seem to cause the same issue.
2. Can we provide a better code pattern for handling these situations? Using `correct_levels` is not obvious. Maybe a function for specifically getting the head without risking changing the scope.
3. Can we avoid copying the whole type of all function arguments just to handle an extremely rare corner case? Can we delay copying until we detect that the expansion would change the scope? Can we avoid copying when there are no local constraints in the environment etc.
4. Can we push the functions that inspect types like this into Ctype? It seems better to avoid heavy type manipulation outside of `ctype.ml`. Anything which explicitly inspects the description of a type is "risky" and it would be nice to contain such code behind an abstraction boundary.

garrigue replied:
* It have made the use of `correct_levels` conditional on `Env.local_constraints`, as it is only meaningful in that case.
* I have looked in `Typecore`, and I can indeed see a number of other suspicious calls to `Env.expand_head` (i.e., for inspection but not strictly required by unification). However, I would first like to find a counter-example before protecting those calls, as it is not even certain that the protection would be efficient.
* As for the need of an abstraction barrier, I agree, but it is currently pretty porous, and I would prefer not to mix this with this PR. I.e., it's probably better to keep semantical changes and refactoring in separate PRs.

lpw25 reviewed (commented):
The core change looks good, but this raises some issues about how careful we are when inspecting the head of a type. I think we need to find some solutions to those issues before we make this change.

garrigue committed:
only use correct_levels when there are local constraints

garrigue commented:
Thanks for the comments, and sorry for the slow answers.

garrigue commented:
This PR requires a bootstrap before merging.

garrigue commented:
For the role of `full_expand`, the best way to understand is to look at what happens if we do not use `may_forget_scope`.
Compared to before this PR, we get many examples where the non-principal and principal versions diverge. So it is hard to say that strictly following the scopes is more correct. There may be a mystery to solve.
```diff
--- test.ml	2021-04-05 09:16:14.000000000 +0900
+++ test.ml.corrected.corrected	2021-04-14 16:08:28.000000000 +0900
@@ -704,6 +704,14 @@
                   ^
 Error: This expression has type < m : a; .. >
        but an expression was expected of type < m : b; .. >
+       This instance of b is ambiguous:
+       it would escape the scope of its equation
+|}, Principal{|
+Line 2, characters 14-15:
+2 |   fun Eq o -> o
+                  ^
+Error: This expression has type < m : a; .. >
+       but an expression was expected of type < m : b; .. >
        Type a is not compatible with type b = a
        This instance of a is ambiguous:
        it would escape the scope of its equation
@@ -717,6 +725,14 @@
                           ^
 Error: This expression has type < m : a; .. >
        but an expression was expected of type < m : b; .. >
+       This instance of b is ambiguous:
+       it would escape the scope of its equation
+|}, Principal{|
+Line 2, characters 22-23:
+2 |   match eq with Eq -> o ;; (* should fail *)
+                          ^
+Error: This expression has type < m : a; .. >
+       but an expression was expected of type < m : b; .. >
        Type a is not compatible with type b = a
        This instance of a is ambiguous:
        it would escape the scope of its equation
@@ -773,6 +789,14 @@
                                                 ^
 Error: This expression has type < m : a; .. >
        but an expression was expected of type < m : b >
+       This instance of b is ambiguous:
+       it would escape the scope of its equation
+|}, Principal{|
+Line 3, characters 44-45:
+3 |     let r : < m : b > = match eq with Eq -> o in (* fail *)
+                                                ^
+Error: This expression has type < m : a; .. >
+       but an expression was expected of type < m : b >
        Type a is not compatible with type b = a
        This instance of a is ambiguous:
        it would escape the scope of its equation
@@ -786,8 +810,7 @@
                   ^
 Error: This expression has type [> `A of a ]
        but an expression was expected of type [> `A of b ]
-       Type a is not compatible with type b = a
-       This instance of a is ambiguous:
+       This instance of b is ambiguous:
        it would escape the scope of its equation
 |}, Principal{|
 Line 2, characters 9-15:
@@ -806,6 +829,14 @@
                           ^
 Error: This expression has type [> `A of a ]
        but an expression was expected of type [> `A of b ]
+       This instance of b is ambiguous:
+       it would escape the scope of its equation
+|}, Principal{|
+Line 2, characters 22-23:
+2 |   match eq with Eq -> v ;; (* should fail *)
+                          ^
+Error: This expression has type [> `A of a ]
+       but an expression was expected of type [> `A of b ]
        Type a is not compatible with type b = a
        This instance of a is ambiguous:
        it would escape the scope of its equation
@@ -1202,6 +1233,14 @@
 5 |     if true then a else b
                             ^
 Error: This expression has type b = int
+       but an expression was expected of type a
+       This instance of a is ambiguous:
+       it would escape the scope of its equation
+|}, Principal{|
+Line 5, characters 24-25:
+5 |     if true then a else b
+                            ^
+Error: This expression has type b = int
        but an expression was expected of type a = int
        Type int is not compatible with type a = int
        This instance of int is ambiguous:
@@ -1218,9 +1257,9 @@
 4 |   if b then x else y
                        ^
 Error: This expression has type b = int
-       but an expression was expected of type a = int
-       Type a = int is not compatible with type a = int
-       This instance of int is ambiguous:
+       but an expression was expected of type a
+       Type a = int is not compatible with type a
+       This instance of a is ambiguous:
        it would escape the scope of its equation
 |}];;
 
@@ -1233,7 +1272,7 @@
 4 |   if b then y else x
                        ^
 Error: This expression has type a = int
-       but an expression was expected of type b = int
-       This instance of int is ambiguous:
+       but an expression was expected of type b
+       This instance of b is ambiguous:
        it would escape the scope of its equation
 |}];;

)
```

lpw25 commented:
I'm still broadly concerned about all the other uses of `expand_head`. Even the uses in `unify` seem to cause spurious ambiguity errors. For example,
```ocaml
# type (_, _) eq = Refl : ('a, 'a) eq;;
type (_, _) eq = Refl : ('a, 'a) eq
# let foo (type a) (x : (a, int) eq) (a : a) p =
    let module M = struct type t = a let x : t = a end in
    if p then a else M.x;;
    val foo : ('a, int) eq -> 'a -> bool -> 'a = <fun>
# let foo (type a) (x : (a, int) eq) (a : a) p =
    let module M = struct type t = a let x : t = a end in
    let Refl = x in
    if p then a else M.x;;
      Line 4, characters 21-24:
4 |     if p then a else M.x;;
                         ^^^
Error: This expression has type M.t = int
       but an expression was expected of type a = int 
       This instance of int is ambiguous:
       it would escape the scope of its equation
```

For cases like the one in `type_argument`, where we don't know if we need the expansion until we look at the expansion, I wonder if we might be able to use `Btype.backtrack` to undo the expansion when it turns out we don't need it. That would be cheaper than copying the type in advance.

garrigue committed:
use backtracking rather than copying in Typecore.type_argument

garrigue commented:
Used backtracking in `type_argument` as suggested. This is indeed a better approach.
I would be better to encapsulate it cleanly; still thinking about a good interface for that.

As for the spurious ambiguity errors, this is a well known problem inherent to our use of type manifests to encode equations.
It may be possible to improve it using an ordering on types (i.e. find the lub of two types in this ordering rather than normalize them), but this is going to be painful for limited benefits (it seems that all examples exposing this problem are rather contrived, since they involve types with several equations).

Edit: actually there is already some support for this in `Ctype.unify`, but it is limited to cases where both sides are local abstract types. Local modules  pose another challenge.

garrigue commented:
Actually I could extend the approach in `unify` to handle your example. See #10348.

lpw25 commented on typing/ctype.ml:
```diff
@@ -1756,8 +1760,20 @@ let enforce_constraints env ty =
-  let ty = repr (expand_head env ty) in
+let full_expand ~may_forget_scope env ty =
+  let ty =
+    if may_forget_scope then
```
I wonder if it would be slightly simpler to pass `may_forget_scope` down to `expand_abbrev_gen` and simply ignore the `type_expansion_scope` when it is true. Perhaps rename the parameter to `ignore_scopes` to make it clearer if you do.

garrigue replied:
I looked into it, but the machinery around `expand_head` is complicated, so that it would be hard to do.
Moreover, I would be concerned by allowing "wrong" handling of potentially shared types (well, we are reporting an error, so this should be fine (either fatal failure, or backtrack in the toplevel), but morally this is still strange).
Since cost is not a problem here, calling `correct_levels` is morally cleaner. We may lose some sharing, but then the goal here is just to show that there is an equation, but we are not allowed to use it.

lpw25 reviewed (approved):
This looks good now that it's using backtracking. I agree that we can leave the other cases of `expand_head` to later.

garrigue committed:
accept error message change

garrigue commented:
Thank you for the reviews.

dra27 commented:
This PR has "broken" the build of Dune - though not on main (because the code in question has been removed). On either the 2.8 or 2.9 branches of Dune, bootstrapping gives:

```
File "src/memo/memo.ml", line 1003, characters 73-74:
Error: This expression has type 'a but an expression was expected of type i
       This instance of i is ambiguous:
       it would escape the scope of its equation
```
the code in question is:
```ocaml
module With_implicit_output = struct
  type ('i, 'o, 'f) t = 'f

  let create (type i o f io) name ?doc ~input ~visibility
      ~output:(module O : Output_simple with type t = o) ~implicit_output
      (typ : (i, o, f) Function.Type.t) (impl : f) =
    let output =
      Output.Simple
        (module struct
          type t = o * io option

          let to_dyn ((o, _io) : t) =
            Dyn.List [ O.to_dyn o; Dyn.String "<implicit output is opaque>" ]
        end)
    in
    match typ with
    | Function.Type.Sync ->
      let memo =
        create name ?doc ~input ~visibility ~output Sync (fun i ->
            Implicit_output.collect_sync implicit_output (fun () -> impl i))
      in
      (fun input ->
         let res, output = exec memo input in
         Implicit_output.produce_opt implicit_output output;
         res
        : f)
    | Function.Type.Async ->
      let memo =
        create name ?doc ~input ~visibility ~output Async (fun i ->
            Implicit_output.collect_async implicit_output (fun () -> impl i))
      in
      (fun input ->
         Fiber.map (exec memo input) ~f:(fun (res, output) ->
             Implicit_output.produce_opt implicit_output output;
             res)
        : f)
```
in the `Type.Sync` match branch.

garrigue commented:
> This PR has "broken" the build of Dune - though not on main (because the code in question has been removed). On either the 2.8 or 2.9 branches of Dune, bootstrapping gives:

Interesting, if not surprising: since it restricts typing, it can indeed break existing programs that use GADTs.

Two solutions:
* either just add an asterisk to the log entry and live with it
* or check that the previous behavior was sound (if not principal), and activate the recursive propagation only in principal mode

The second option has the advantage of limiting breakage, but the soundness is not completely immediate.
The reason scoping is needed for soundness is to ensure that equations do not leak physically outside of a branch, through the sharing introduced by unification.
However unification shares only nodes above a node corresponding to an equation, but nodes below it are unified with an instance of the equation, and as such are not shared, so that the previous behavior should already be sufficient for soundness alone.

dra27 commented:
Is it worth going to opam to see? I rebased this branch onto 4.12 and dune gets the same issue, but I have this failure in the testsuite... I'm not sure if it's that I've made a mistake transferring the code or if there's something else on trunk which should be transplanted to 4.12 just for this exercise?

```diff
--- test.ml     2021-04-23 21:17:56.637397287 +0100
+++ test.ml.corrected.corrected 2021-04-23 21:18:57.502332327 +0100
@@ -631,8 +631,8 @@
 Line 3, characters 46-47:
 3 |     let module M = struct type b = a let z = (y : b) end
                                                   ^
-Error: This expression has type a = int
-       but an expression was expected of type b = int
+Error: This expression has type 'a but an expression was expected of type
+         a = int
        This instance of int is ambiguous:
        it would escape the scope of its equation
 |}];;

```

garrigue commented:
@dra27 Thanks for checking!
The change in error message is well in the margin of acceptability, it comes probably from another recent change in trunk (actually I remember seeing this change recently, albeit not exactly where), but I don't think there is any real dependency for this PR.

It would be great indeed if you could test opam.
I did not expect many failures, but dune is well known to heavily exercise GADTs.
Depending on the result, we may choose the option of only enabling this change in principal mode, but this would require a new PR.

kit-ty-kate commented:
This PR seems to have also broken `parsexp` (though the patch had already been upstreamed during the first 4.12 alpha due to a similar change that has been removed: https://github.com/janestreet/parsexp/commit/4a4f08e08a9be78db9dd38821e80318cbe47dd96)

However this is not the only failure. From most used to least used:
```
#=== ERROR while compiling async_rpc_kernel.v0.14.0 ===========================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/async_rpc_kernel.v0.14.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p async_rpc_kernel -j 31
# exit-code            1
# env-file             ~/.opam/log/async_rpc_kernel-7-15223c.env
# output-file          ~/.opam/log/async_rpc_kernel-7-15223c.out
### output ###
#     ocamlopt src/.async_rpc_kernel.objs/native/async_rpc_kernel__Implementations.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12.1+pr10277/bin/ocamlopt.opt -w -40 -g -I src/.async_rpc_kernel.objs/byte -I src/.async_rpc_kernel.objs/native -I /home/opam/.opam/4.12.1+pr10277/lib/async_kernel -I /home/opam/.opam/4.12.1+pr10277/lib/async_kernel/persistent_connection_kernel -I /home/opam/.opam/4.12.1+pr10277/lib/base -I /home/opam/.opam/4.12.1+pr10277/lib/base/base_internalhash_types -I /home/opam/.opam/4.12.1+pr10277/lib/base/caml -I /home/opam/.opam/4.12.1+pr10277/lib/base/md5 -I /home/opam/.opam/4.12.1+pr10277/lib/base/shadow_stdlib -I /home/opam/.opam/4.12.1+pr10277/lib/base_bigstring -I /home/opam/.opam/4.12.1+pr10277/lib/base_quickcheck -I /home/opam/.opam/4.12.1+pr10277/lib/bin_prot -I /home/opam/.opam/4.12.1+pr10277/lib/bin_prot/shape -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/base_for_tests -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/moption -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/pairing_heap -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/sexp_hidden_in_test -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/thread_pool_cpu_affinity -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/thread_safe_queue -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/timing_wheel -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/tuple_pool -I /home/opam/.opam/4.12.1+pr10277/lib/core_kernel/uopt -I /home/opam/.opam/4.12.1+pr10277/lib/fieldslib -I /home/opam/.opam/4.12.1+pr10277/lib/jane-street-headers -I /home/opam/.opam/4.12.1+pr10277/lib/parsexp -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_assert/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_bench/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_enumerate/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/collector -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/common -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/config -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/config_types -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_here/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_inline_test/config -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_inline_test/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_module_timer/runtime -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/protocol_version_header -I /home/opam/.opam/4.12.1+pr10277/lib/sexplib -I /home/opam/.opam/4.12.1+pr10277/lib/sexplib0 -I /home/opam/.opam/4.12.1+pr10277/lib/splittable_random -I /home/opam/.opam/4.12.1+pr10277/lib/stdio -I /home/opam/.opam/4.12.1+pr10277/lib/time_now -I /home/opam/.opam/4.12.1+pr10277/lib/typerep -I /home/opam/.opam/4.12.1+pr10277/lib/variantslib -intf-suffix .ml -no-alias-deps -open Async_rpc_kernel__ -o src/.async_rpc_kernel.objs/native/async_rpc_kernel__Implementations.cmx -c -impl src/implementations.pp.ml)
# File "src/implementations.ml", line 522, characters 39-45:
# 522 |         | Implementation.F.Deferred -> result
#                                              ^^^^^^
# Error: This expression has type
#          $Rpc_expert_'result =
#            Async_rpc_kernel__Implementation.Expert.implementation_result
#            Async_kernel__Types.Deferred.t
#        but an expression was expected of type
#          'a Async_kernel.Monitor.Deferred.t 
#        This instance of 'a is ambiguous:
#        it would escape the scope of its equation
```
```
#=== ERROR while compiling x509.0.13.0 ========================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/x509.0.13.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p x509 -j 47
# exit-code            1
# env-file             ~/.opam/log/x509-7-6c34a9.env
# output-file          ~/.opam/log/x509-7-6c34a9.out
### output ###
#       ocamlc lib/.x509.objs/byte/x509__General_name.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12.1+pr10277/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.x509.objs/byte -I lib/.x509.objs/public_cmi -I /home/opam/.opam/4.12.1+pr10277/lib/asn1-combinators -I /home/opam/.opam/4.12.1+pr10277/lib/astring -I /home/opam/.opam/4.12.1+pr10277/lib/base/caml -I /home/opam/.opam/4.12.1+pr10277/lib/base64 -I /home/opam/.opam/4.12.1+pr10277/lib/bigarray-compat -I /home/opam/.opam/4.12.1+pr10277/lib/bytes -I /home/opam/.opam/4.12.1+pr10277/lib/cstruct -I /home/opam/.opam/4.12.1+pr10277/lib/domain-name -I /home/opam/.opam/4.12.1+pr10277/lib/eqaf -I /home/opam/.opam/4.12.1+pr10277/lib/eqaf/bigstring -I /home/opam/.opam/4.12.1+pr10277/lib/eqaf/cstruct -I /home/opam/.opam/4.12.1+pr10277/lib/fmt -I /home/opam/.opam/4.12.1+pr10277/lib/gmap -I /home/opam/.opam/4.12.1+pr10277/lib/logs -I /home/opam/.opam/4.12.1+pr10277/lib/mirage-crypto -I /home/opam/.opam/4.12.1+pr10277/lib/mirage-crypto-ec -I /home/opam/.opam/4.12.1+pr10277/lib/mirage-crypto-pk -I /home/opam/.opam/4.12.1+pr10277/lib/mirage-crypto-rng -I /home/opam/.opam/4.12.1+pr10277/lib/parsexp -I /home/opam/.opam/4.12.1+pr10277/lib/pbkdf -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ptime -I /home/opam/.opam/4.12.1+pr10277/lib/result -I /home/opam/.opam/4.12.1+pr10277/lib/rresult -I /home/opam/.opam/4.12.1+pr10277/lib/seq -I /home/opam/.opam/4.12.1+pr10277/lib/sexplib -I /home/opam/.opam/4.12.1+pr10277/lib/sexplib0 -I /home/opam/.opam/4.12.1+pr10277/lib/stdlib-shims -I /home/opam/.opam/4.12.1+pr10277/lib/zarith -no-alias-deps -open X509__ -o lib/.x509.objs/byte/x509__General_name.cmo -c -impl lib/general_name.ml)
# File "lib/general_name.ml", line 121, characters 28-46:
# 121 |       | Other oid, [ x ] -> `C1 (`C1 (oid, x))
#                                   ^^^^^^^^^^^^^^^^^^
# Error: This expression has type [> `C1 of [> `C1 of Asn.oid * string ] ]
#        but an expression was expected of type 'a 
#        This instance of string is ambiguous:
#        it would escape the scope of its equation
```
```
#=== ERROR while compiling linol.0.2 ==========================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/linol.0.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build @install -p linol -j 31
# exit-code            1
# env-file             ~/.opam/log/linol-7-68a0f1.env
# output-file          ~/.opam/log/linol-7-68a0f1.out
### output ###
#       ocamlc src/.linol.objs/byte/linol__Server.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12.1+pr10277/bin/ocamlc.opt -w -40 -warn-error -a+8 -g -bin-annot -I src/.linol.objs/byte -I /home/opam/.opam/4.12.1+pr10277/lib/biniou -I /home/opam/.opam/4.12.1+pr10277/lib/bytes -I /home/opam/.opam/4.12.1+pr10277/lib/containers -I /home/opam/.opam/4.12.1+pr10277/lib/containers/monomorphic -I /home/opam/.opam/4.12.1+pr10277/lib/csexp -I /home/opam/.opam/4.12.1+pr10277/lib/easy-format -I /home/opam/.opam/4.12.1+pr10277/lib/jsonrpc -I /home/opam/.opam/4.12.1+pr10277/lib/lsp -I /home/opam/.opam/4.12.1+pr10277/lib/lsp/stdune -I /home/opam/.opam/4.12.1+pr10277/lib/ocaml/threads -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_yojson_conv_lib -I /home/opam/.opam/4.12.1+pr10277/lib/result -I /home/opam/.opam/4.12.1+pr10277/lib/seq -I /home/opam/.opam/4.12.1+pr10277/lib/stdlib-shims -I /home/opam/.opam/4.12.1+pr10277/lib/uchar -I /home/opam/.opam/4.12.1+pr10277/lib/uutf -I /home/opam/.opam/4.12.1+pr10277/lib/yojson -no-alias-deps -open Linol__ -o src/.linol.objs/byte/linol__Server.cmo -c -impl src/server.ml)
# File "src/server.ml", line 122, characters 10-88:
# 122 |           self#on_req_completion ~uri:textDocument.uri ~pos:position ~ctx:context doc_st
#                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This expression has type
#          [ `CompletionList of Lsp.Types.CompletionList.t
#          | `List of Lsp.Types.CompletionItem.t list ] option IO.t
#        but an expression was expected of type r IO.t
#        Type
#          [ `CompletionList of Lsp.Types.CompletionList.t
#          | `List of Lsp.Types.CompletionItem.t list ] option
#        is not compatible with type
#          r =
#            [ `CompletionList of Lsp__.Types.CompletionList.t
#            | `List of Lsp__.Types.CompletionItem.t list ] option  
#        This instance of [ `CompletionList of Lsp.Types.CompletionList.t
#                         | `List of Lsp.Types.CompletionItem.t list ] is ambiguous:
#        it would escape the scope of its equation
```
```
#=== ERROR while compiling amqp-client.2.2.2 ==================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/amqp-client.2.2.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p amqp-client -j 31
# exit-code            1
# env-file             ~/.opam/log/amqp-client-7-80ed6f.env
# output-file          ~/.opam/log/amqp-client-7-80ed6f.out
### output ###
#       ocamlc lib/.amqp_client_lib.objs/byte/amqp_client_lib__Protocol.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12.1+pr10277/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.amqp_client_lib.objs/byte -I /home/opam/.opam/4.12.1+pr10277/lib/bytes -I /home/opam/.opam/4.12.1+pr10277/lib/ocplib-endian -no-alias-deps -open Amqp_client_lib -o lib/.amqp_client_lib.objs/byte/amqp_client_lib__Protocol.cmo -c -impl lib/protocol.ml)
# File "lib/protocol.ml", line 63, characters 4-22:
# 63 |     read_table_value t
#          ^^^^^^^^^^^^^^^^^^
# Error: This expression has type (string * 'a) list
#        but an expression was expected of type
#          a = (string * Amqp_client_lib.Types.value) list 
#        This instance of 'a is ambiguous:
#        it would escape the scope of its equation
```
```
#=== ERROR while compiling eliom.7.0.0 ========================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/eliom.7.0.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build make
# exit-code            2
# env-file             ~/.opam/log/eliom-8-7fa474.env
# output-file          ~/.opam/log/eliom-8-7fa474.out
### output ###
# ocamlfind ocamlc -c -g -keep-locs -thread -package js_of_ocaml-ppx_deriving_json -package js_of_ocaml-ppx -package xml-light -package react,lwt_react,js_of_ocaml -package lwt,ocsigenserver,ocsigenserver.ext,tyxml.functor -package js_of_ocaml-ppx_deriving_json,lwt_ppx -w +A-4-6-7-9-27-37-39-40-42-44-48 -I /home/opam/.opam/4.12.1+pr10277/lib/js_of_ocaml -I src/lib -I src/lib/server -o src/lib/server/eliom_parameter_base.cmo src/lib/server/eliom_parameter_base.ml
# + ocamlfind ocamlc -c -g -keep-locs -thread -package js_of_ocaml-ppx_deriving_json -package js_of_ocaml-ppx -package xml-light -package react,lwt_react,js_of_ocaml -package lwt,ocsigenserver,ocsigenserver.ext,tyxml.functor -package js_of_ocaml-ppx_deriving_json,lwt_ppx -w +A-4-6-7-9-27-37-39-40-42-44-48 -I /home/opam/.opam/4.12.1+pr10277/lib/js_of_ocaml -I src/lib -I src/lib/server -o src/lib/server/eliom_parameter_base.cmo src/lib/server/eliom_parameter_base.ml
# File "src/lib/eliom_parameter_base.shared.ml", line 288, characters 28-78:
# Error: This expression has type (string * Eliommod_parameters.field) list
#        but an expression was expected of type
#          (string * Eliommod_parameters.field) list 
#        This instance of string is ambiguous:
#        it would escape the scope of its equation
# Command exited with code 2.
# make: *** [Makefile:14: native] Error 10
```
```
#=== ERROR while compiling bechamel-js.0.1.0 ==================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/bechamel-js.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p bechamel-js -j 31
# exit-code            1
# env-file             ~/.opam/log/bechamel-js-11-dc1629.env
# output-file          ~/.opam/log/bechamel-js-11-dc1629.out
### output ###
#       ocamlc lib/js/.bechamel_js.objs/byte/bechamel_js.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12.1+pr10277/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/js/.bechamel_js.objs/byte -I /home/opam/.opam/4.12.1+pr10277/lib/bechamel -I /home/opam/.opam/4.12.1+pr10277/lib/fmt -I /home/opam/.opam/4.12.1+pr10277/lib/jsonm -I /home/opam/.opam/4.12.1+pr10277/lib/ocplib-json-typed -I /home/opam/.opam/4.12.1+pr10277/lib/result -I /home/opam/.opam/4.12.1+pr10277/lib/rresult -no-alias-deps -open Bechamel_js__ -o lib/js/.bechamel_js.objs/byte/bechamel_js.cmo -c -impl lib/js/bechamel_js.ml)
# File "lib/js/bechamel_js.ml", line 132, characters 9-16:
# 132 |         (ref buf, ref off, ref len)
#                ^^^^^^^
# Error: This expression has type bytes ref
#        but an expression was expected of type 'a 
#        This instance of bytes is ambiguous:
#        it would escape the scope of its equation
```
```
#=== ERROR while compiling elpi.1.13.1 ========================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/elpi.1.13.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build make build DUNE_OPTS=-p elpi -j 71
# exit-code            2
# env-file             ~/.opam/log/elpi-7-fcf3e8.env
# output-file          ~/.opam/log/elpi-7-fcf3e8.out
### output ###
# File "src/runtime.ml", line 1934, characters 62-67:
# Error: This expression has type 'a but an expression was expected of type int 
#        This instance of int is ambiguous:
#        it would escape the scope of its equation
# make: *** [Makefile:31: build] Error 1
```
```
#=== ERROR while compiling pacomb.1.0 =========================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/pacomb.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p pacomb -j 47
# exit-code            1
# env-file             ~/.opam/log/pacomb-7-cea15d.env
# output-file          ~/.opam/log/pacomb-7-cea15d.out
### output ###
# (cd _build/default && /home/opam/.opam/4.12.1+pr10277/bin/ocamlopt.opt -w -40 -w -9 -g -I lib/.pacomb.objs/byte -I lib/.pacomb.objs/native -I lib/.pacomb.objs/public_cmi -I /home/opam/.opam/4.12.1+pr10277/lib/stdlib-shims -intf-suffix .ml -no-alias-deps -open Pacomb -o lib/.pacomb.objs/native/pacomb__Container.cmx -c -impl lib/container.pp.ml)
# File "lib/container.ml", line 140, characters 15-16:
# 140 |       | Y -> f v;
#                      ^
# Error: This expression has type ($E_'a, 'a) elt = a
#        but an expression was expected of type a 
#        This instance of 'a is ambiguous:
#        it would escape the scope of its equation
```
```
#=== ERROR while compiling pythonlib.v0.14.0 ==================================#
# context              2.1.0~beta4 | linux/x86_64 | ocaml-variants.4.12.1+pr10277 | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.1+pr10277/.opam-switch/build/pythonlib.v0.14.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p pythonlib -j 47
# exit-code            1
# env-file             ~/.opam/log/pythonlib-9-8bd75d.env
# output-file          ~/.opam/log/pythonlib-9-8bd75d.out
### output ###
# (cd _build/default && /home/opam/.opam/4.12.1+pr10277/bin/ocamlopt.opt -w -40 -g -I src/.python_lib.objs/byte -I src/.python_lib.objs/native -I /home/opam/.opam/4.12.1+pr10277/lib/base -I /home/opam/.opam/4.12.1+pr10277/lib/base/base_internalhash_types -I /home/opam/.opam/4.12.1+pr10277/lib/base/caml -I /home/opam/.opam/4.12.1+pr10277/lib/base/shadow_stdlib -I /home/opam/.opam/4.12.1+pr10277/lib/jane-street-headers -I /home/opam/.opam/4.12.1+pr10277/lib/ocaml/compiler-libs -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_enumerate/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/collector -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/common -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/config -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_expect/config_types -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_here/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_inline_test/config -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_inline_test/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_python/runtime -I /home/opam/.opam/4.12.1+pr10277/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.12.1+pr10277/lib/pyml -I /home/opam/.opam/4.12.1+pr10277/lib/seq -I /home/opam/.opam/4.12.1+pr10277/lib/sexplib0 -I /home/opam/.opam/4.12.1+pr10277/lib/stdcompat -I /home/opam/.opam/4.12.1+pr10277/lib/stdio -I /home/opam/.opam/4.12.1+pr10277/lib/time_now -I /home/opam/.opam/4.12.1+pr10277/lib/typerep -intf-suffix .ml -no-alias-deps -open Python_lib__ -o src/.python_lib.objs/native/python_lib__Defunc.cmx -c -impl src/defunc.pp.ml)
# File "src/defunc.ml", line 166, characters 8-12:
# 166 |         args, { state with pos = total_args_len; after_star_args = true }
#               ^^^^
# Error: This expression has type Python_lib__.Import.pyobject list
#        but an expression was expected of type
#          a = Python_lib__.Import.pyobject list 
#        This instance of Pytypes.pyobject is ambiguous:
#        it would escape the scope of its equation
```
The first two have a number of packages depending on them that I couldn't test yet, so some of them might also fail

Octachron commented:
I have small example (extracted from odoc) which is failing after this change that seems strange
```ocaml
type _ t = A: int option t
let fail (type a) x (A: a t) = (Some x:a)
```
The failure seems even stranger since this variant (which is tested in the testsuite) works
```ocaml
type _ v = A: int v
let ok (type a) x (A: a v) = (x:a)
```

garrigue commented:
@Octachron This failure is not strange, it is exactly the goal of this change.
In the `fail` case, one has to use the equation on `a` to find that `Some x` has type `int option`, and `x` has type `int`.
So the equation is indeed leaking (in this example there is no other possible type for `x`, but one can devise other examples where there could be).
In the `ok` case there is no need to use the equation, so `x` has just type `a`.
Your examples are interesting in that they show that ambivalence and non-ambiguity are two independent concepts: `fail` is ambivalent but not ambiguous (there is only one possible type), while `ok` is non-ambivalent, but ambiguous in that `ok` could have another type (if we were to use the equation).

@kit-ty-kate Thanks for your reports.
I did not investigate the code, but this seems to be cases this change is intended to catch, and they should be fixable by adding type annotations.
However, I'm starting to think that enabling this change only in `-principal` mode would limit the disruption. I'll try to come up with a PR soon.

garrigue commented:
@kit-ty-kate I have created a PR #10383 disabling the change for non-principal mode.
This works as expected on the test suite, removing errors introduced by this PR, keeping them only in principal mode.
Could you check it on opam?

kit-ty-kate commented:
Sure. I'll post the results tomorrow morning if nothing has gone wrong.

kit-ty-kate commented:
I've posted the results directly in the new PR (TL;DR it looks good)

gasche commented on typing/ctype.ml:
```diff
@@ -1652,10 +1652,14 @@ let expand_head_once env ty =
+  | Unify _ ->
+      Btype.backtrack snap;
+      cleanup_abbrev ();
+      false
```
Why did this change introduce a call to `cleanup_abbrev`? My understanding is that `Btype.backtrack` already calls `cleanup_abbrev`.

gasche commented on typing/ctype.ml:
```diff
@@ -1671,7 +1675,7 @@ let try_expand_safe env ty =
   try try_expand_once env ty
   with Unify _ ->
-    Btype.backtrack snap; raise Cannot_expand
+    Btype.backtrack snap; cleanup_abbrev (); raise Cannot_expand
```
same question here on `cleanup_abbrev` after `Btype.backtrack`.

gasche reviewed (commented):
@garrigue this PR introduced calls to `cleanup_abbrev` that seem unnecessary to me. Can you confirm?
